Add hot reloading of the engine and fix game hot-reloading

JDuchniewicz opened the pull request:
Moves the engine + renderer to be a separate DLL called `runtime`. This DLL is loaded and then the engine loads the game DLL. This way in `hot-reload` build types the developer can quickly iterate without restarting the whole game and losing the state. It also heavily improves the hot-reload time due to using `mold` linker and choosing better compilation flags for this profile.

The standalone executable communicates with the runtime via versioned ABI and mismatches cause the exec to terminate. This ABI is expected to be fairly stable (hence versioning has one digit) as we don't expect a lot of changes in the windowing/input passing between standalone and the runtime world.

The ABI between Runtime and Game is soft (meaning they still communicate via Rust types and Rust internals). This can blow when the changes include changing API between Game/Engine. Hot-reload is not a production feature and that's why we accept that (rather than having a gazillion C-like functions that remove the ease of ECS usage). 

TODO: 
- [ ] State saving - serde of the entire world's state
- [ ] Detecting changes that would cause incompatibility and prompting for a restart?
- [x] Merge with the new way of handling `winit` as already done on `main` branch

JDuchniewicz committed:
WiP allowing for faster hot-reloading by restructuring how we compile things.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Further reduce hot-reload time. Build alongside pill_standalone to maintain proper type mappings.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Improving reload speed.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Plug in runtime hot-reloading.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Achieve 0.56s hot-reload time on Linux when changing one file.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
React to changes in engine/pill_core|engine|renderer/src/* directories.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
hot-reloading: Introduce pill_abi and pill_runtime crates. Make the engine + renderer a separate runtime DLL.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Naming fixes.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Fix path mess.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Sanity restored. Not overwriting existing libs on reload.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Cleanup old hot-reloaded libs on rebuild.

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

MattSzymonski commented:
Damn, looks great! 🚀

MattSzymonski commented:
I have found few issues:
1. Took me a while to discover that I should use `-c hot-reload` :P
2. Without this flag standalone crashes with `configuration property "WINDOW_TITLE" not found`. This is because game's `res` folder is not copied to `/build/dev/data` by the launcher.
3. Second hot reload of the game crashes for me with `Failed to remove file: D:\Programming\Pill-Engine\examples\floating_pills\build\hot-reload\data\pill_game_loaded_0.dll` Wrong dll generation set in the name?

Info: On my machine game hot reload takes 2.3846086s (where 1.8396574s is game project build)

MattSzymonski commented:
I think it is time to split standalone and launcher into multiple files. It is getting hard to read.

JDuchniewicz commented:
>     1. Took me a while to discover that I should use `-c hot-reload` :P
Right, will update the README. We now have 3 build options, with `hot-reload` being the dev-only way of doing things

>     2. Without this flag standalone crashes with `configuration property "WINDOW_TITLE" not found`. This is because game's `res` folder is not copied to `/build/dev/data` by the launcher.
This can also be fixed. In general, I think we should discuss the directory tree, I like how it is done right now but I am open to discussions to make it more straightforward (avoid copying stuff around).

>     3. Second hot reload of the game crashes for me with `Failed to remove file: D:\Programming\Pill-Engine\examples\floating_pills\build\hot-reload\data\pill_game_loaded_0.dll` Wrong dll generation set in the name?
This is done so that we cleanup old hot-reload DLLs before the new hot-reload run. Maybe it's permissions on Windows? The DLL name matches? It works fine on Linux though.

> Info: On my machine game hot reload takes 2.3846086s (where 1.8396574s is game project build)
So the hot-reload of engine takes around this amount for me as well - changing one file deeper in the dependency tree. The hot-reload of the game is much faster. Especially if you do incremental changes - the subsequent hot-reload builds are much faster. It depends on the complexity of the change one made.

The hot-reload time should remain constant, it's only the build time that is varying and over that we have no further control except tweaking `cargo` flags or using `ccache` to hold temporary build artifacts.

JDuchniewicz commented:
> I think it is time to split standalone and launcher into multiple files. It is getting hard to read.

I also believe this should be done. I propose it is done in a separate PR to reduce the amount of noise. Also, I am thinking of adding the serde changes in a separate PR for the same reason. Let me know your thoughts.

JDuchniewicz committed:
docs: add hot-reload info to README

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
rebase onto main manually

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
launcher: fix wrong path handling

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz commented:
Rebased and addressed the issues mentioned above. Sorry it took so long, got distracted by other projects. Let's get up to speed with that! Please take a look @MattSzymonski @sp0lsh

JDuchniewicz committed:
ci: Fix according to new layout

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>

JDuchniewicz committed:
Merge branch 'main' into hot_reloading

JDuchniewicz commented:
Also, somehow I scored over 2700 FPS on the FloatingPills demo for the release on 2.5K resolution (!?). Seems like we are finally getting somewhere!

JDuchniewicz committed:
clippy: fix lints

Signed-off-by: Jakub Duchniewicz <j.duchniewicz@gmail.com>
